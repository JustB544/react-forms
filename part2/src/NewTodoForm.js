import React, {useState} from "react";

function NewTodoForm({createTodo}) {
    const INITIAL_STATE = {task: ""};
    const [formData, setFormData] = useState(INITIAL_STATE);
    function handleSubmit(evt){
        evt.preventDefault();
        createTodo(formData);
        setFormData(INITIAL_STATE);
    }

    function handleChange(evt){
        const {name, value} = evt.target;
        setFormData(fData => ({...fData, [name]: value}));
    }
    return (
        <form style={{display: "flex", flexDirection: "column", width: "fit-content"}} onSubmit={handleSubmit}>
            <div>
                <label htmlFor="task">Task: </label>
                <input name="task" type="text" value={formData.task} placeholder="New task" onChange={handleChange}/>
            </div>
            <button>Create</button>
        </form>
    )
}

export default NewTodoForm;