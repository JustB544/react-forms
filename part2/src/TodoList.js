import React, {useState} from "react";
import Todo from "./Todo";
import {v4 as uuid} from 'uuid';
import NewTodoForm from "./NewTodoForm";

function TodoList(props) {
    const [todos, setTodos] = useState(props.initial || {});

    function deleteTodo(id){
        setTodos(({[id]: undefined, ...t}) => t);
    }

    function createTodo({ task }){
        setTodos(t => ({...t, [uuid()]: { task }}));
    }

    return (
        <div>
            { Object.keys(todos).map(t => <Todo 
            key={t} 
            id={t} 
            task={todos[t].task}
            deleteTodo={evt => deleteTodo(t)}/>) }
            <NewTodoForm createTodo={(todo, evt) => createTodo(todo)}/>
        </div>
    );
}

export default TodoList;