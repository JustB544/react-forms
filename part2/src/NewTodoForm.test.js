import { render, fireEvent } from "@testing-library/react";
import NewTodoForm from "./NewTodoForm";
import TodoList from "./TodoList";

it("renders", () => {
    render(<NewTodoForm />);
});
  
it("matches snapshot", () => {
    const {asFragment} = render(<NewTodoForm />);
    expect(asFragment()).toMatchSnapshot();
});

it("creates a todo", () => {
    const {container} = render(<TodoList/>);
    const button = container.querySelector("form button");
    fireEvent.click(button);
    expect(container.querySelector(".Todo")).toBeInTheDocument();
});