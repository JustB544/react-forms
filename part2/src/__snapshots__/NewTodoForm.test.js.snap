// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form
    style="display: flex; flex-direction: column;"
  >
    <div>
      <label
        for="task"
      >
        Task: 
      </label>
      <input
        name="task"
        placeholder="New task"
        type="text"
        value=""
      />
    </div>
    <button>
      Create
    </button>
  </form>
</DocumentFragment>
`;
